File: sig_deferring_sigpending.py
#!/bin/env python
import sys
from pysigset import *
from time import sleep
from signal import *
from posix_timers import *

if __name__ == "__main__":
    st=10 if len(sys.argv)!=2 else int(sys.argv[1])
    print "sleep for %ds in each round\n" %(st)

    s=SIGSET()
    saved_s=SIGSET()
    pending_s=SIGSET()
    #sigemptyset(s)
    sigaddset(s,SIGINT)
    n=0
    MICRO=1.0/1000000.0
    while True:
        sigprocmask(SIG_BLOCK,s,saved_s)
        bt=clock_gettime(CLOCK_MONOTONIC)
        sleep(st)
        et=clock_gettime(CLOCK_MONOTONIC)
        print "sleep:%f" %(et-bt)
        ++n
        sigpending(pending_s)
        if sigismember(pending_s,SIGINT):
            print "Gotten a SIGINT!!\n"
            break
        else:
            pass

File: sig_deferring_sigpending.c
#include<unistd.h>
#include<sys/signalfd.h>
#include<stdio.h>
#include<stdlib.h>
#include<signal.h>
#include<time.h>

int main(int argc,char**argv){
	sigset_t set,saved_set,pending_set;
	sigemptyset(&set);
	sigaddset(&set,SIGINT);
	int fd=signalfd(-1,&set,SFD_NONBLOCK);
	struct signalfd_siginfo siginfo;

	int n=0;
	int rc;
	struct timespec begin,end,elapse,remain;
	elapse.tv_sec=argc!=2?10:atoi(argv[1]);
	elapse.tv_nsec=0;
	const double NANO=1.0/1000000000;
	printf("sleep for %d in each round\n",elapse.tv_sec);
	while(1){
		sigprocmask(SIG_BLOCK,&set,&saved_set);

		clock_gettime(CLOCK_MONOTONIC,&begin);
		nanosleep(&elapse,&remain);
		clock_gettime(CLOCK_MONOTONIC,&end);
		double ta=remain.tv_sec+remain.tv_nsec*NANO;
		double tb=end.tv_sec-begin.tv_sec
				 +(end.tv_nsec-begin.tv_nsec >0? (end.tv_nsec-begin.tv_nsec)*NANO
						                       : (1+end.tv_nsec-begin.tv_nsec)*NANO-1
				  );		
		printf("%d: sleep:%.3f remain:%.3f\n",n,tb,ta);
		++n;
		sigemptyset(&pending_set);
		sigpending(&pending_set);

		if (sigismember(&pending_set,SIGINT)){
			printf("Gotten a SIGINT!!!\n");
			break;
		}
	}
}
